### PR TITLE
Add support for integration tests

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
This adds support for integration tests.

First there are only two basic test "suites", create folder and upload file.
Both have two test cases, which try to create folder / upload file in one existing folder and in an non-existing folder.
(with the later one [uploading a file into a non-existing folder] we could have avoid 3.0.3).

It spins up a server with latest master, which as a user test, where we can do testing.

CodeCov takes reports from unit and integration tests and shows a summary.

After all this is only a very basic/tiny step to introduce testing:
- write unit tests
- write integration tests
- write UI tests

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

File: .codecov.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    - drone.nextcloud.com
+
+coverage:
+  precision: 2
+  round: down
+  range: "70...100"
+
+comment:
+  layout: "header, diff, changes, uncovered, tree"
+  behavior: default
```
Comment (mario) [inline .codecov.yml]: new line.

**Comment by arundhati24** [comment]
@tobiasKaminsky How about using Mockito for this? :)

**Comment by tobiasKaminsky** [comment]
@arundhati24 yes, I will also use Mockito for unit tests :+1: 
Starting next week I want to add tests for any new PR prior merging them.

This PR should enable us to use unit tests and integration tests with code coverage and a nice way of getting infos if tests fail.

**Comment by arundhati24** [comment]
Sure @tobiasKaminsky :+1:

**Comment by tobiasKaminsky** [comment]
This is ready for review.
I would like to have this soon in, so we can see if and how real testing is working.
We can then redefine it later, if needed.
First I think we should start with writing unit tests, as @ardevd already did for the activity PR.

**Comment by AndyScherzinger** [comment]
:+1: looks good to me

File: .codecov.yml (hunk 1 above)
Comment (AndyScherzinger) [reply .codecov.yml]: has been added :)

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger I would like to do this in a step after merging of this, so we can see if and how this works out in the wild, prior to put too much effort in it.

**Comment by AndyScherzinger** [comment]
> I would like to do this in a step after merging of this, so we can see if and how this works out in the wild, prior to put too much effort in it.

perfectly fine by me

@mario can you have a look/review? :)

**Comment by AndyScherzinger** [comment]
rebased to latest master pinging @mario @ardevd for second review/approval. :)

**Comment by AndyScherzinger** [comment]
cc @mario @ardevd for second review/approval.

**Comment by ardevd** [comment]
I use Gitlab's CI system myself so I don't really have any experience with the setup you guys have going. But from a pure Android perspective, it looks good to me.

**Comment by mario** [comment]
@ardevd you indicate the approval by writing 👍  as the first char.

File: scripts/uploadReport.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    cd build/reports/androidTests/connected/flavors/GPLAY
+else 
+    cd build/reports/tests/testGplayDebugUnitTest
+fi
+
+find . -type d -exec curl -u $USER:$PASS -X MKCOL $URL/$ID/$(echo {} | sed s#\./##) \;
+find . -type f -exec curl -u $USER:$PASS -X PUT $URL/$ID/$(echo {} | sed s#\./##) --upload-file {} \;
+
+echo "Uploaded failing tests to https://nextcloud.kaminsky.me/index.php/s/XqY52jBr9ZYfDiz -> $ID" 
+exit 1
```
Comment (mario) [inline scripts/uploadReport.sh]: New line.

File: src/androidTest/java/com/owncloud/android/screenshots/ScreenshotsIT.java [hunk 3]
```diff
@@ -1,131 +1,131 @@
...
-        onData(PreferenceMatchers.withTitle(R.string.prefs_category_more)).perform(ViewActions.scrollTo());
-
-        Thread.sleep(1000);
-        
-        Screengrab.screenshot("06_davdroid");
-
-        Assert.assertTrue(true); // if we reach this, everything is ok
-    }
-}
+//package com.owncloud.android.screenshots;
```
Comment (mario) [inline src/androidTest/java/com/owncloud/android/screenshots/ScreenshotsIT.java]: Remove the file if not used?

File: src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java [hunk 4]
```diff
@@ -69,51 +65,51 @@ public void checkPreconditions() {
...
-        Context context = InstrumentationRegistry.getContext();
-        final Intent intent = context.getPackageManager()
-                .getLaunchIntentForPackage(OWNCLOUD_APP_PACKAGE);
-        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
-        context.startActivity(intent);
-
-        // Wait for the app to appear
-        mDevice.wait(Until.hasObject(By.pkg(OWNCLOUD_APP_PACKAGE).depth(0)), LAUNCH_TIMEOUT);
-    }
+//    @Test
```
Comment (mario) [inline src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java]: Remove things not used.

File: src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java [hunk 5]
```diff
@@ -69,51 +65,51 @@ public void checkPreconditions() {
     /**
      * Start owncloud app
```
Comment (mario) [inline src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java]: nextcloud

**Comment by mario** [comment]
Few comments.

File: src/androidTest/java/com/owncloud/android/screenshots/ScreenshotsIT.java (hunk 3 above)
Comment (tobiasKaminsky) [reply src/androidTest/java/com/owncloud/android/screenshots/ScreenshotsIT.java]: This is needed for automated screenshots, but should not be used for automated testing.
Therefore I disabled the tests, maybe you have a better idea?

File: src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java (hunk 4 above)
Comment (tobiasKaminsky) [reply src/androidTest/java/com/owncloud/android/uiautomator/InitialTest.java]: These are "old" tests from oC, which I plan to reactivate or delete if they are not working. But first I wanted to have integration tests working on drone.

**Comment by tobiasKaminsky** [comment]
Great, for me it is fine to merge.
We would then have coverage reports, which might fail, but at least it is a step forward to remind us that we need test cases.

**Comment by mario** [comment]
@tobiasKaminsky can you comment on why we have so much code with // in front though? Shouldn't we remove this?

**Comment by tobiasKaminsky** [comment]
ScreenshotIT:
> This is needed for automated screenshots, but should not be used for automated testing.
Therefore I disabled the tests, maybe you have a better idea? 

and

InitalTest:
> These are "old" tests from oC, which I plan to reactivate or delete if they are not working. But first I wanted to have integration tests working on drone.

**Comment by mario** [comment]
Uncomment, and move to a different directory? Would be cleaner for me.

**Comment by AndyScherzinger** [comment]
:+1: still fine by me :)

**Comment by AndyScherzinger** [comment]
@mario 
> Uncomment, and move to a different directory? Would be cleaner for me.
DONE :)

So is the PR fine by you?

**Comment by mario** [comment]
👍

**Pull request merged**